Prusti Server Mode

juliand665 opened the pull request:
As part of my Bachelor's Thesis (under Federico Poli), I'm implementing a "server mode" for Prusti in which it keeps around a launched JVM as well as caching verifier instances to get rid of startup delays in verification and take load off clients.

This is a WIP draft (not ready for review yet), mostly for tracking purposes.

juliand665 committed:
macOS support in makefile

juliand665 committed:
fix benchmark compilation errors

juliand665 committed:
more portable makefile

juliand665 committed:
standardize makefile formatting

juliand665 committed:
add serde crate & standardize versions

juliand665 committed:
implement Serialize & Deserialize for VIR AST

juliand665 committed:
extract Program & its encoding logic

that actually extracted really well

juliand665 committed:
implement Serialize & Deserialize for Program

juliand665 committed:
move optimization logic to dedicated place

this also lessens the dependency on prusti_interface::config; to be completely removed soon

juliand665 committed:
make verification output serializable

juliand665 committed:
Merge branch 'master' of gitlab.inf.ethz.ch:OU-PMUELLER/prusti-dev into prusti-server
update

juliand665 committed:
comment cleanup

juliand665 committed:
basic tarpc service definition

heavily work-in-progress

juliand665 committed:
standardize spelling of "optimize"

US spelling

juliand665 committed:
basic server-side implementation

juliand665 committed:
config boilerplate reduction

juliand665 committed:
move utils

juliand665 committed:
unify timing logic

juliand665 committed:
split tarpc service and verification service

this way plays much nicer with the dependencies and encapsulates stuff more meaningfully

juliand665 committed:
formatting

juliand665 committed:
implement thread-based verifier logic

juliand665 committed:
actually implemented connecting to server

entirely untested of course, but this makes sense to me

juliand665 committed:
switch from callbacks to futures

juliand665 committed:
remove disused makefile

juliand665 committed:
extract common utilities across prusti to prusti-common

this clarifies prusti-interface's role as dealing with rustc, rather than being general-purpose utilities.

also some formatting changed via format-on-save (stable rustfmt)

juliand665 committed:
extract more utilities to driver_utils

prusti-filter and prusti itself sure have a lot in common

juliand665 committed:
move config & report to prusti-common

fpoli commented on prusti-common/README.md:
```diff
@@ -0,0 +1,3 @@
+# prusti-common
+
+Common utilities used across the Prusti project.
```
I would add "compiler-agnostic" (utilities), as the idea is to never use rustc as a library in this crate.

fpoli commented on prusti-common/src/lib.rs:
```diff
@@ -0,0 +1,30 @@
+}
+
+/// Runs a given function, timing and logging (info-level) how long it took, returning the function's result.
+pub fn run_timed<F: FnOnce() -> T, T>(desc: &'static str, task: F) -> T {
```
Does it also work if `desc: &'static str` is declared as `desc: &str`?

fpoli commented on prusti-interface/src/utils.rs:
```diff
@@ -4,7 +4,7 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-//! Various helper functions for working with `mir::Place`.
+//! Various helpers functions for working with `mir::Place`.
```
Could you check the correct spelling?

juliand665 replied:
oops, no idea how that slipped through!

fpoli commented on prusti-server/src/lib.rs:
```diff
@@ -0,0 +1,78 @@
+    fn get_or_create_thread(&mut self, backend: VerificationBackend) -> &VerifierThread {
+        if !self.threads.contains_key(&backend) {
+            let thread = VerifierThread::new(self.verifier_builder.clone(), backend);
+            self.threads.insert(backend, thread);
```
You still need to implement a limit on the thread pool size, right?

juliand665 replied:
That's correct; I'll add a `TODO` note.

fpoli commented on prusti-server/src/service.rs:
```diff
@@ -0,0 +1,84 @@
+        Self { client }
+    }
+
+    pub fn new_fake() -> Self {
```
Could you add brief a rustdoc explaining the `_fake`?

juliand665 replied:
This whole method is a little questionable; I've renamed it and expanded in the comments/docs

fpoli commented on prusti-server/src/verifier_thread.rs:
```diff
@@ -0,0 +1,61 @@
+                let ast_factory = context.new_ast_factory();
+                while let Ok(request) = request_receiver.recv() {
+                    let viper_program = request.program.to_viper(&ast_factory);
+                    let result = verifier.verify(viper_program);
```
It would be nice to measure and log timings for the calls to the JVM.

juliand665 replied:
Indeed. In the newest commits, this will be taken care of.

fpoli commented on prusti-viper/src/encoder/vir/cfg/method.rs:
```diff
@@ -12,30 +12,35 @@ use uuid::Uuid;
-#[derive(Debug, Clone)]
-pub struct CfgMethod {
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CfgMethod { // TODO: Should labels and reserved_labels be ignored?
```
Regarding the TODO: the best thing here would be to
1. rename `CfgMethod` to `CfgMethodBuilder`, and remove the serialize/deserialize from it
2. add a new, much simpler, `CfgMethod` (serializable, without `labels`, `reserved_labels`, ...)
3. add a `fn build(self) -> CfgMethod` to `CfgMethodBuilder`
4. move any `to_viper` method from `CfgMethodBuilder` to `CfgMethod`

If we don't do this, then skipping label fields is ok as a temporary solution (if they are not used in the conversion to Viper).

fpoli commented on prusti-viper/src/encoder/vir/program.rs:
```diff
@@ -0,0 +1,94 @@
+}
+
+impl Program {
+    pub fn optimized(mut self) -> Self {
```
Could you move
```rust
impl Program {
    pub fn optimized(...) -> Self { ...}
}
```
to the `mod.rs` of the `optimizations` module? Such that `vir` doesn't use stuff from `optimizations`.

fpoli commented on prusti-viper/src/encoder/vir/program.rs:
```diff
@@ -0,0 +1,94 @@
+        self
+    }
+
+    pub fn to_viper<'v>(self, ast: &AstFactory<'v>) -> viper::Program<'v> {
```
Could you move this method to the `to_viper` module, implementing the `ToViper` trait for `Program`?

juliand665 replied:
Oh yeah, for some reason I got confused a while back when this idea first came up. Because the code uses `into_iter()`, which requires ownership, I thought I couldn't easily implement `ToViper`, but it turns out it works just fine with regular old `iter()`, so I've implemented the trait now.

fpoli commented on prusti-viper/src/verification_service.rs:
```diff
@@ -0,0 +1,12 @@
+pub enum ViperBackendConfig {
+    Silicon(Vec<String>),
+    Carbon(Vec<String>),
+}
```
Could you add a brief rustdoc that explains what is a `ViperBackendConfig` and what are the `Vec<String>`?

fpoli reviewed (commented):
I added some comments.

juliand665 committed:
encapsulate verification logic

so as to be usable directly on the client

juliand665 committed:
unify viper verifier usage

now all the configuration uses ViperBackendConfig, and the prusti verifier uses all the infrastructure we've built up rather than managing its own verifier

juliand665 committed:
minor improvements

fixes for issues that were brought up in code review

juliand665 committed:
clean up, remove hacks, clarify TODOs

juliand665 committed:
move consistency error handling

now panics in prusti-viper (client-side) rather than viper (server-side)

juliand665 committed:
WIP

juliand665 committed:
update

juliand665 committed:
fix compilation errors

juliand665 committed:
add copyright notices

juliand665 committed:
improve run_timed ergonomics

now prints two messages (pre + post), taking a description of the stage rather than a completed action as argument

juliand665 committed:
use run_timed in viper module

juliand665 committed:
tiny cleanup

juliand665 committed:
fix failing test

can't believe i missed this; took way too long to localize…

juliand665 committed:
change viper test expectations

panic -> result case
also added a quick way of running just them

juliand665 committed:
use limited LRU cache for verifier threads

thanks to the lru crate, this was spectacularly easy (once i had figured out which version i could use with our compiler version)

juliand665 committed:
server address config option

this is the heart of making the client connect to the server

juliand665 committed:
tiny cleanup

juliand665 committed:
use server for unit tests

this, uhh, single-handedly halves the time they all take.

juliand665 committed:
deduplicate some test logic

had this lying around on a branch and just now found out how to express the trait bound (for<'v> F: FnOnce(&AstFactory<'v>) -> Stmt<'v>,), so here goes

juliand665 committed:
add bodyless predicate

this lets us move the hardcoding logic from to_viper to encoder

juliand665 committed:
initial version of server executable + driver

can't test this right now because prusti-viper depends on rustc. VIR/to_viper will move to prusti-common soon

juliand665 committed:
merge imports when formatting rustfmt

juliand665 committed:
remove dependency of viper on prusti-common

didn't really make sense, since prusti-common is for prusti.

juliand665 committed:
update from master

a huge bunch of merge conflicts; almost definitely doesn't compile right now

juliand665 committed:
reapply optimization spelling change

juliand665 committed:
fix remaining compiler issues

everything went better than expected

juliand665 committed:
update from master

juliand665 committed:
improve timing API

run_timed! -> Stopwatch
decided to make it finish on drop to ease including return values in timing, so now there is no way to use it invalidly

juliand665 committed:
move driver utils to own crate

so i can link against them with the server executable

juliand665 committed:
move verifier builder & context to prusti-common

the server needs those, of course

fpoli commented on prusti-common/src/config.rs:
```diff
@@ -184,6 +178,19 @@ pub fn report_support_status() -> bool {
+    read_setting("SERVER_MAX_STORED_VERIFIERS")
+}
+
+/// When set, Prusti will connect to this server and use it for its verification backend (i.e. the things using the JVM/Viper).
+/// Set to "MOCK" to run the server off-thread, effectively mocking connecting to a server without having to start it up separately.
```
Is `<ip>:<port>` a valid `SERVER_ADDRESS`?

fpoli commented on prusti-server/src/verifier_runner.rs:
```diff
@@ -0,0 +1,74 @@
+
+        stopwatch.start_next_section("verification");
+        self.verifier.verify(viper_program)
+    }
```
Could you measure and log the duration of `to_viper`, `dump` and `verify` with the `Stopwatch`?

fpoli commented on prusti-viper/src/verifier.rs:
```diff
@@ -228,148 +231,28 @@ impl<'v, 'r, 'a, 'tcx> Verifier<'v, 'r, 'a, 'tcx> {
-        );
+        let mut program = self.encoder.get_viper_program();
+        if config::simplify_encoding() {
+            program = program.optimized();
```
Could you measure `optimized` with a `Stopwatch`?

fpoli commented on prusti-viper/src/verifier.rs:
```diff
@@ -228,148 +231,28 @@ impl<'v, 'r, 'a, 'tcx> Verifier<'v, 'r, 'a, 'tcx> {
-            _ => vec![],
+            viper::VerificationResult::ConsistencyErrors(errors) => {
+                errors.iter().for_each(|e| error!("{}", e));
+                panic!("Consistency errors. The encoded Viper program is incorrect.");
```
Instead of `panic`, could you try using `PrustiError::internal("...", DUMMY_SP).emit();`, returning `Failure`?

juliand665 committed:
cleanup

juliand665 committed:
fix remaining issues with standalone server executable

juliand665 committed:
move driver utilities to new prusti-launch crate

juliand665 committed:
move more verification logic to prusti-viper/verifier

also renamed prusti-common/verifier

juliand665 committed:
tiny cleanup

juliand665 committed:
add prusti-server back to prusti dependencies

need this for unit tests after all

juliand665 committed:
move handling of some config flags to client

juliand665 committed:
update from master

juliand665 committed:
fix bad merge

this took unjustifiably long to narrow down…

juliand665 committed:
rename start_next_section to start_next

much shorter this way, yet still clear

juliand665 committed:
make stopwatch use a domain prefix

also added some documentation around the place

juliand665 committed:
remove some unnecessary extern crates

juliand665 committed:
report verifier startup times

juliand665 committed:
change default server log level to info

no need to be careful since we're not outputting through rustc

juliand665 committed:
update from master

# Conflicts:
#	prusti-common/src/config.rs
#	prusti-common/src/vir/ast/common.rs
#	prusti-viper/src/encoder/encoder.rs
#	prusti-viper/src/encoder/procedure_encoder.rs
#	prusti-viper/src/verifier.rs
#	prusti/src/compiler_calls.rs
#	prusti/tests/test.rs

juliand665 committed:
encode viper domains in program

juliand665 committed:
update from master

# Conflicts:
#	Cargo.lock
#	prusti-common/src/vir/ast/common.rs
#	prusti-viper/Cargo.toml
